src/module.c

```diff
@@ -442,6 +444,29 @@ char *RM_Strdup(const char *str) {
     return zstrdup(str);
 }
 
+void RM_ProcessEventsWhileBlocked(RedisModuleCtx *ctx) {
+    long long now = getMonotonicUs();
+    long long elapsed = elapsedMs(ctx->start_time);
+
+    if (ctx->start_time == 0) {
+        ctx->start_time = ctx->next_event = now;
+        return;
+    }
+
+    if (elapsed < server.script_time_limit)
+        return;
+
+    if (now >= ctx->next_event) {
+        server.busy_job = 1;
+        blockingOperationStarts();
```

Comment:
`blockingOperationStarts` should only be called once (on the first time we get here.
and then `blockingOperationEnds` should be called when the context is destroyed.

Reply:
@oranagra  blockingOperationEnds should happen in moduleFreeContext

Reply:
i think so.. but maybe it's more complicated than that when there are nested contexts?

i.e. the two typical cases i was thinking of are simple commands called by RedisModuleCommandDispatcher (will respond with -BUSY), and about contexts during the `rdb_load` callback (should respond with -LOADING.

but maybe there are also cases to think about with regards to other more complex scenarios.
i think you can use moduleFreeContext for now, and let's ask @guybe7 and @MeirShpilraien to have a look.

Reply:
we need to remember that this function can be reentrant: a module command module.foo calls this API, while another module command module.bar is allow-busy and is waiting to be processed
when it is finally processed here, it also calls this module API

Reply:
@oranagra, wdyt ?

Reply:
i guess we should do the unblocking when in `moduleFreeContext` when `server.module_ctx_nesting` goes back to 0.

p.s. maybe we should forbid calling these APIs from threads (`REDISMODULE_CTX_THREAD_SAFE`)?
or maybe there's actually an advantage to let modules handle events in long blocking GIL locks from threads?
(i feel sick by just raising that idea).
@MeirShpilraien ?

Reply:
@oranagra I think there is a bug advantage of allowing it from a `ThreadSafeCtx` maybe on `ThreadSafeCtx` we can call `blockingOperationStarts` and `blockingOperationEnd` from the lock and unlock functions? Also notice that `blockingOperationStarts` and `blockingOperationEnd` already handled nesting so maybe there is no need to check `server.module_ctx_nesting`

Reply:
LOL "bug advantage". Exactly my thoughts. 

Anyway, just calling the starts / stops methods isn't gonna do anything. And we certainly don't wanna call the new (processEvents) API automatically (it'll start rejecting commands). 
And we can allow calling the new API in thread safe contexts, and after a while it'll start processing events.